File: backend/src/roteiro.ts
export interface RoteiroUpdate {
  path: (string | number)[];
  value: unknown;
}

export function applyUpdates(data: unknown, updates: RoteiroUpdate[]): unknown {
  const proximo = structuredClone(data) as Record<string | number, unknown>;
  for (const { path, value } of updates) {
    if (path.length === 0) continue;
    let alvo: Record<string | number, unknown> = proximo;
    for (let i = 0; i < path.length - 1; i++) {
      const chave = path[i];
      const proximoAlvo = alvo[chave];
      if (typeof proximoAlvo !== "object" || proximoAlvo === null) {
        alvo[chave] = typeof path[i + 1] === "number" ? [] : {};
      }
      alvo = alvo[chave] as Record<string | number, unknown>;
    }
    alvo[path[path.length - 1]] = value;
  }
  return proximo;
}

interface EspinhaNo {
  id: string;
  tipo: string;
  ordem?: number;
  conteudo: string;
  status: string;
  conecta_assets: string[];
  excluido?: boolean;
}

/**
 * Calcula um `ordem` (a chave de ordenação de exibição das complicações,
 * puramente numérica — não precisa ser inteira) que encaixa uma
 * complicação na posição de exibição `posicao` (1-based) entre as
 * complicações de `visiveis` (que NÃO inclui a que está sendo posicionada).
 * `posicao` omitida ou além do fim vai pro final da lista. Insere entre a
 * vizinha anterior e a seguinte usando um valor fracionário, pra nunca
 * precisar renumerar mais ninguém.
 */
function calcularOrdemParaPosicao(visiveis: EspinhaNo[], posicao: number | undefined, maiorOrdem: number): number {
  if (posicao === undefined || posicao > visiveis.length) {
    return maiorOrdem + 1;
  }
  const alvo = Math.max(1, posicao);
  const anterior = visiveis[alvo - 2];
  const seguinte = visiveis[alvo - 1];
  const ordemAnterior = anterior?.ordem ?? 0;
  const ordemSeguinte = seguinte?.ordem ?? ordemAnterior + 2;
  return (ordemAnterior + ordemSeguinte) / 2;
}

/**
 * `posicao` (1-based) é a posição de exibição que a complicação nova deve
 * ocupar entre as complicações visíveis atuais — omitida, ela vai pro fim
 * (mesmo comportamento de sempre). O índice real dela no array `espinha`
 * continua sendo sempre o último elemento (nunca insere no meio do array
 * de verdade — ver o porquê na fatia do nó repetível); o que muda é só o
 * campo `ordem`, a chave de ordenação pra exibição.
 */
export function adicionarComplicacao(data: unknown, posicao?: number): unknown {
  const proximo = structuredClone(data) as { espinha: EspinhaNo[] };
  const todasComplicacoes = proximo.espinha.filter((no) => no.tipo === "complicacao");
  const proximoId = todasComplicacoes.length + 1;

  const visiveis = todasComplicacoes.filter((no) => !no.excluido).sort((a, b) => (a.ordem ?? 0) - (b.ordem ?? 0));
  const maiorOrdem = todasComplicacoes.reduce((max, no) => Math.max(max, no.ordem ?? 0), 0);
  const ordem = calcularOrdemParaPosicao(visiveis, posicao, maiorOrdem);

  proximo.espinha.push({
    id: `complicacao_${proximoId}`,
    tipo: "complicacao",
    ordem,
    conteudo: "",
    status: "vazio",
    conecta_assets: ["antagonista", "protagonista"],
  });
  return proximo;
}

/**
 * Reposiciona uma complicação que já existe pro seu novo lugar de exibição
 * (1-based), identificada pelo `id` (nunca por índice — o índice real no
 * array nunca muda). Mesmo cálculo de `ordem` fracionário do
 * `adicionarComplicacao`, mas contra a lista de vizinhas SEM a própria
 * complicação que está sendo movida.
 */
export function reordenarComplicacao(data: unknown, idComplicacao: string, novaPosicao: number): unknown {
  const proximo = structuredClone(data) as { espinha: EspinhaNo[] };
  const no = proximo.espinha.find((n) => n.id === idComplicacao);
  if (!no || no.tipo !== "complicacao" || no.excluido) {
    throw new Error(`complicação "${idComplicacao}" não encontrada ou não está visível`);
  }

  const todasComplicacoes = proximo.espinha.filter((n) => n.tipo === "complicacao");
  const maiorOrdem = todasComplicacoes.reduce((max, n) => Math.max(max, n.ordem ?? 0), 0);
  const visiveisSemAlvo = todasComplicacoes
    .filter((n) => !n.excluido && n.id !== idComplicacao)
    .sort((a, b) => (a.ordem ?? 0) - (b.ordem ?? 0));

  no.ordem = calcularOrdemParaPosicao(visiveisSemAlvo, novaPosicao, maiorOrdem);
  return proximo;
}

export function excluirComplicacao(data: unknown, indice: number): unknown {
  const proximo = structuredClone(data) as { espinha: EspinhaNo[] };
  const no = proximo.espinha[indice];
  if (!no || no.tipo !== "complicacao") {
    throw new Error("índice não corresponde a uma complicação");
  }
  no.excluido = true;
  return proximo;
}

export function moverComplicacao(data: unknown, indice: number, direcao: "cima" | "baixo"): unknown {
  const proximo = structuredClone(data) as { espinha: EspinhaNo[] };
  const no = proximo.espinha[indice];
  if (!no || no.tipo !== "complicacao" || no.excluido) {
    throw new Error("índice não corresponde a uma complicação visível");
  }

  const visiveis = proximo.espinha
    .map((n, i) => ({ n, i }))
    .filter(({ n }) => n.tipo === "complicacao" && !n.excluido)
    .sort((a, b) => (a.n.ordem ?? 0) - (b.n.ordem ?? 0));
  const posicao = visiveis.findIndex(({ i }) => i === indice);
  const vizinho = direcao === "cima" ? visiveis[posicao - 1] : visiveis[posicao + 1];
  if (!vizinho) {
    throw new Error("não há complicação vizinha nessa direção");
  }

  const ordemDoNo = no.ordem;
  no.ordem = vizinho.n.ordem;
  vizinho.n.ordem = ordemDoNo;
  return proximo;
}

export function espinhaVazia() {
  return [
    { id: "incidente_incitante", tipo: "no_fixo", conteudo: "", status: "vazio", conecta_assets: ["protagonista", "mundo"] },
    { id: "complicacao_1", tipo: "complicacao", ordem: 1, conteudo: "", status: "vazio", conecta_assets: ["antagonista", "protagonista"] },
    { id: "crise", tipo: "no_fixo", conteudo: "", status: "vazio", conecta_assets: ["protagonista"] },
    { id: "climax", tipo: "no_fixo", conteudo: "", status: "vazio", conecta_assets: ["protagonista", "ideia_controladora"] },
    { id: "resolucao", tipo: "no_fixo", conteudo: "", status: "vazio", conecta_assets: [] },
  ];
}

export function roteiroMckeeVazio() {
  return {
    template: "mckee",
    titulo: "",
    fase_atual: "A",
    espinha: espinhaVazia(),
    assets: {
      protagonistas: [
        {
          id: "prot_1",
          want: "",
          need: "",
          aposta: "",
          caracterizacao: "",
          carater_verdadeiro: "",
          arco: "",
          pov: "",
          status: "vazio",
        },
      ],
      antagonista: {
        niveis: [],
        fonte_oposicao: "",
        logica_interna: "",
        avatar: "",
        poder_relativo: "",
        pontos_contato: [],
        status: "vazio",
      },
      ideia_controladora: { valor: "", causa: "", contraideia: "", status: "vazio" },
      mundo: { epoca: "", local: "", regras_custo: "", status: "vazio" },
      genero: { generos: [], promessa: "", status: "vazio" },
      elenco_notas: { texto_livre: "" },
    },
  };
}

File: agent/roteiro.py
"""Schema e mutações do roteiro McKee — porte de backend/src/roteiro.ts (legado)."""


def calcular_ordem_para_posicao(visiveis: list[dict], posicao: int | None, maior_ordem: float) -> float:
    """Encaixa uma complicação na posição de exibição `posicao` (1-based) com
    `ordem` fracionária — insere entre as vizinhas sem renumerar ninguém.
    Omitida ou além do fim, vai pro final."""
    if posicao is None or posicao > len(visiveis):
        return maior_ordem + 1
    alvo = max(1, posicao)
    anterior = visiveis[alvo - 2] if alvo >= 2 else None
    seguinte = visiveis[alvo - 1] if alvo - 1 < len(visiveis) else None
    ordem_anterior = (anterior or {}).get("ordem", 0) or 0
    ordem_seguinte = seguinte.get("ordem", ordem_anterior + 2) if seguinte else ordem_anterior + 2
    return (ordem_anterior + ordem_seguinte) / 2


def adicionar_complicacao(data: dict, conteudo: str, posicao: int | None = None) -> dict:
    """Como no legado, o nó novo vai sempre pro FIM do array `espinha` (o
    índice real nunca muda); a posição de exibição é só o campo `ordem`.
    Diferença deliberada do laboratório: recebe `conteudo` direto (no legado
    a ação criava vazio e o conteúdo vinha por proposta)."""
    todas = [no for no in data["espinha"] if no["tipo"] == "complicacao"]
    proximo_id = len(todas) + 1
    visiveis = sorted(
        (no for no in todas if not no.get("excluido")),
        key=lambda no: no.get("ordem", 0) or 0,
    )
    maior_ordem = max((no.get("ordem", 0) or 0 for no in todas), default=0)
    ordem = calcular_ordem_para_posicao(visiveis, posicao, maior_ordem)
    novo = {
        "id": f"complicacao_{proximo_id}",
        "tipo": "complicacao",
        "ordem": ordem,
        "conteudo": conteudo,
        "status": "rascunho",
        "conecta_assets": ["antagonista", "protagonista"],
    }
    # Cópia rasa basta: só o array espinha muda (append); os nós existentes
    # e os assets não são mutados — não precisa de deepcopy do roteiro todo.
    return {**data, "espinha": [*data["espinha"], novo]}


# Espelho da config de cartões do frontend (web/src/lib/cartoes.ts) — os
# campos de texto que o agente enxerga e pode propor via propor_campo.
CAMPOS_CARTOES: dict[str, list[str]] = {
    "protagonistas": ["want", "need", "aposta"],
    "antagonista": ["fonte_oposicao", "logica_interna", "avatar", "poder_relativo"],
    "ideia_controladora": ["valor", "causa", "contraideia"],
    "mundo": ["epoca", "local", "regras_custo"],
    "genero": ["promessa"],
}


def _asset(data: dict, asset: str) -> dict:
    # Defensivo: o roteiro pode vir do estado enviado pelo CLIENTE a cada
    # turno — uma forma malformada não pode derrubar o run inteiro.
    alvo = (data.get("assets") or {}).get(asset)
    if asset == "protagonistas":
        alvo = alvo[0] if alvo else None
    return alvo if isinstance(alvo, dict) else {}


def resumo_assets(data: dict) -> str:
    linhas = []
    for asset, campos in CAMPOS_CARTOES.items():
        alvo = _asset(data, asset)
        linhas.append(f"[{asset}] (status: {alvo.get('status', '?')})")
        linhas.extend(f"- {c}: {alvo.get(c) or '(vazio)'}" for c in campos)
    return "\n".join(linhas)


def resumo_espinha(data: dict) -> str:
    """Resumo compacto da espinha em ordem de exibição, com as posições
    1-based das complicações — é o que o modelo usa pra escolher `posicao`."""
    def chave(no: dict) -> float:
        if no["tipo"] == "complicacao":
            return 1 + (no.get("ordem", 0) or 0) / 1000
        return {"incidente_incitante": 0, "crise": 2, "climax": 3, "resolucao": 4}.get(no["id"], 99)

    visiveis = sorted(
        (no for no in data.get("espinha") or [] if not no.get("excluido")),
        key=chave,
    )
    linhas = []
    pos_complicacao = 0
    for no in visiveis:
        if no["tipo"] == "complicacao":
            pos_complicacao += 1
            rotulo = f"Complicação (posição {pos_complicacao}, id {no['id']})"
        else:
            rotulo = no["id"]
        conteudo = no.get("conteudo") or "(vazio)"
        linhas.append(f"- {rotulo}: {conteudo}")
    return "\n".join(linhas)


def espinha_vazia() -> list[dict]:
    return [
        {"id": "incidente_incitante", "tipo": "no_fixo", "conteudo": "", "status": "vazio", "conecta_assets": ["protagonista", "mundo"]},
        {"id": "complicacao_1", "tipo": "complicacao", "ordem": 1, "conteudo": "", "status": "vazio", "conecta_assets": ["antagonista", "protagonista"]},
        {"id": "crise", "tipo": "no_fixo", "conteudo": "", "status": "vazio", "conecta_assets": ["protagonista"]},
        {"id": "climax", "tipo": "no_fixo", "conteudo": "", "status": "vazio", "conecta_assets": ["protagonista", "ideia_controladora"]},
        {"id": "resolucao", "tipo": "no_fixo", "conteudo": "", "status": "vazio", "conecta_assets": []},
    ]


def roteiro_mckee_vazio() -> dict:
    return {
        "template": "mckee",
        "titulo": "",
        "fase_atual": "A",
        "espinha": espinha_vazia(),
        "assets": {
            "protagonistas": [
                {
                    "id": "prot_1",
                    "want": "",
                    "need": "",
                    "aposta": "",
                    "caracterizacao": "",
                    "carater_verdadeiro": "",
                    "arco": "",
                    "pov": "",
                    "status": "vazio",
                }
            ],
            "antagonista": {
                "niveis": [],
                "fonte_oposicao": "",
                "logica_interna": "",
                "avatar": "",
                "poder_relativo": "",
                "pontos_contato": [],
                "status": "vazio",
            },
            "ideia_controladora": {"valor": "", "causa": "", "contraideia": "", "status": "vazio"},
            "mundo": {"epoca": "", "local": "", "regras_custo": "", "status": "vazio"},
            "genero": {"generos": [], "promessa": "", "status": "vazio"},
            "elenco_notas": {"texto_livre": ""},
        },
    }
